File: src/strands/agent/conversation_manager/sliding_window_conversation_manager.py
"""Sliding window conversation history management."""

import logging
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ...agent.agent import Agent

from ...hooks import BeforeModelCallEvent, HookRegistry
from ...types.content import ContentBlock, Messages
from ...types.exceptions import ContextWindowOverflowException
from ...types.tools import ToolResultContent
from .conversation_manager import ConversationManager

logger = logging.getLogger(__name__)

_PRESERVE_CHARS = 200


class SlidingWindowConversationManager(ConversationManager):
    """Implements a sliding window strategy for managing conversation history.

    This class handles the logic of maintaining a conversation window that preserves tool usage pairs and avoids
    invalid window states.

    When truncation is enabled (the default), large tool results are partially truncated, preserving the first
    and last 200 characters, and image blocks inside tool results are replaced with descriptive text placeholders.
    Truncation targets the oldest tool results first so the most relevant recent context is preserved as long
    as possible.

    Supports proactive management during agent loop execution via the per_turn parameter.
    """

    def __init__(
        self,
        window_size: int = 40,
        should_truncate_results: bool = True,
        *,
        per_turn: bool | int = False,
    ):
        """Initialize the sliding window conversation manager.

        Args:
            window_size: Maximum number of messages to keep in the agent's history.
                Defaults to 40 messages.
            should_truncate_results: Truncate tool results when a message is too large for the model's context window
            per_turn: Controls when to apply message management during agent execution.
                - False (default): Only apply management at the end (default behavior)
                - True: Apply management before every model call
                - int (e.g., 3): Apply management before every N model calls

                When to use per_turn: If your agent performs many tool operations in loops
                (e.g., web browsing with frequent screenshots), enable per_turn to proactively
                manage message history and prevent the agent loop from slowing down. Start with
                per_turn=True and adjust to a specific frequency (e.g., per_turn=5) if needed
                for performance tuning.

        Raises:
            ValueError: If per_turn is 0 or a negative integer.
        """
        if isinstance(per_turn, int) and not isinstance(per_turn, bool) and per_turn <= 0:
            raise ValueError(f"per_turn must be a positive integer, True, or False, got {per_turn}")

        super().__init__()

        self.window_size = window_size
        self.should_truncate_results = should_truncate_results
        self.per_turn = per_turn
        self._model_call_count = 0

    def register_hooks(self, registry: "HookRegistry", **kwargs: Any) -> None:
        """Register hook callbacks for per-turn conversation management.

        Args:
            registry: The hook registry to register callbacks with.
            **kwargs: Additional keyword arguments for future extensibility.
        """
        super().register_hooks(registry, **kwargs)

        # Always register the callback - per_turn check happens in the callback
        registry.add_callback(BeforeModelCallEvent, self._on_before_model_call)

    def _on_before_model_call(self, event: BeforeModelCallEvent) -> None:
        """Handle before model call event for per-turn management.

        This callback is invoked before each model call. It tracks the model call count and applies message management
        based on the per_turn configuration.

        Args:
            event: The before model call event containing the agent and model execution details.
        """
        # Check if per_turn is enabled
        if self.per_turn is False:
            return

        self._model_call_count += 1

        # Determine if we should apply management
        should_apply = False
        if self.per_turn is True:
            should_apply = True
        elif isinstance(self.per_turn, int) and self.per_turn > 0:
            should_apply = self._model_call_count % self.per_turn == 0

        if should_apply:
            logger.debug(
                "model_call_count=<%d>, per_turn=<%s> | applying per-turn conversation management",
                self._model_call_count,
                self.per_turn,
            )
            self.apply_management(event.agent)

    def get_state(self) -> dict[str, Any]:
        """Get the current state of the conversation manager.

        Returns:
            Dictionary containing the manager's state, including model call count for per-turn tracking.
        """
        state = super().get_state()
        state["model_call_count"] = self._model_call_count
        return state

    def restore_from_session(self, state: dict[str, Any]) -> list | None:
        """Restore the conversation manager's state from a session.

        Args:
            state: Previous state of the conversation manager

        Returns:
            Optional list of messages to prepend to the agent's messages.
        """
        result = super().restore_from_session(state)
        self._model_call_count = state.get("model_call_count", 0)
        return result

    def apply_management(self, agent: "Agent", **kwargs: Any) -> None:
        """Apply the sliding window to the agent's messages array to maintain a manageable history size.

        This method is called after every event loop cycle to apply a sliding window if the message count
        exceeds the window size.

        Args:
            agent: The agent whose messages will be managed.
                This list is modified in-place.
            **kwargs: Additional keyword arguments for future extensibility.
        """
        messages = agent.messages

        if len(messages) <= self.window_size:
            logger.debug(
                "message_count=<%s>, window_size=<%s> | skipping context reduction", len(messages), self.window_size
            )
            return
        self.reduce_context(agent)

    def reduce_context(self, agent: "Agent", e: Exception | None = None, **kwargs: Any) -> None:
        """Trim the oldest messages to reduce the conversation context size.

        The method handles special cases where trimming the messages leads to:
         - toolResult with no corresponding toolUse
         - toolUse with no corresponding toolResult

        Args:
            agent: The agent whose messages will be reduce.
                This list is modified in-place.
            e: The exception that triggered the context reduction, if any.
            **kwargs: Additional keyword arguments for future extensibility.

        Raises:
            ContextWindowOverflowException: If the context cannot be reduced further.
                Such as when the conversation is already minimal or when tool result messages cannot be properly
                converted.
        """
        messages = agent.messages

        # Try to truncate the tool result first
        oldest_message_idx_with_tool_results = self._find_oldest_message_with_tool_results(messages)
        if oldest_message_idx_with_tool_results is not None and self.should_truncate_results:
            logger.debug(
                "message_index=<%s> | found message with tool results at index", oldest_message_idx_with_tool_results
            )
            results_truncated = self._truncate_tool_results(messages, oldest_message_idx_with_tool_results)
            if results_truncated:
                logger.debug("message_index=<%s> | tool results truncated", oldest_message_idx_with_tool_results)
                return

        # Try to trim index id when tool result cannot be truncated anymore
        # If the number of messages is less than the window_size, then we default to 2, otherwise, trim to window size
        trim_index = 2 if len(messages) <= self.window_size else len(messages) - self.window_size

        # Find the next valid trim_index
        while trim_index < len(messages):
            if (
                # Oldest message cannot be a toolResult because it needs a toolUse preceding it
                any("toolResult" in content for content in messages[trim_index]["content"])
                or (
                    # Oldest message can be a toolUse only if a toolResult immediately follows it.
                    any("toolUse" in content for content in messages[trim_index]["content"])
                    and trim_index + 1 < len(messages)
                    and not any("toolResult" in content for content in messages[trim_index + 1]["content"])
                )
            ):
                trim_index += 1
            else:
                break
        else:
            # If we didn't find a valid trim_index, then we throw
            raise ContextWindowOverflowException("Unable to trim conversation context!") from e

        # trim_index represents the number of messages being removed from the agents messages array
        self.removed_message_count += trim_index

        # Overwrite message history
        messages[:] = messages[trim_index:]

    def _truncate_tool_results(self, messages: Messages, msg_idx: int) -> bool:
        """Truncate tool results and replace image blocks in a message to reduce context size.

        For text blocks within tool results, all blocks are partially truncated unless they
        have already been truncated. The first and last _PRESERVE_CHARS characters are kept,
        and the removed middle is replaced with a notice indicating how many characters were
        removed. The tool result status is not changed.

        Image blocks nested inside tool result content are replaced with a short descriptive placeholder.

        Args:
            messages: The conversation message history.
            msg_idx: Index of the message containing tool results to truncate.

        Returns:
            True if any changes were made to the message, False otherwise.
        """
        if msg_idx >= len(messages) or msg_idx < 0:
            return False

        def _image_placeholder(image_block: Any) -> str:
            source: Any = image_block.get("source", {})
            media_type = image_block.get("format", "unknown")
            data = source.get("bytes", b"")
            return f"[image: {media_type}, {len(data) if data else 0} bytes]"

        message = messages[msg_idx]
        changes_made = False
        new_content: list[ContentBlock] = []

        for content in message.get("content", []):
            if "toolResult" in content:
                tool_result: Any = content["toolResult"]
                tool_result_items = tool_result.get("content", [])
                new_items: list[ToolResultContent] = []
                item_changed = False

                for item in tool_result_items:
                    # Replace image items nested inside toolResult content
                    if "image" in item:
                        new_items.append({"text": _image_placeholder(item["image"])})
                        item_changed = True
                        continue

                    # Partially truncate text items that have not already been truncated
                    if "text" in item:
                        text = item["text"]
                        truncation_marker = "... [truncated:"
                        if truncation_marker not in text and len(text) > 2 * _PRESERVE_CHARS:
                            prefix = text[:_PRESERVE_CHARS]
                            suffix = text[-_PRESERVE_CHARS:]
                            removed = len(text) - 2 * _PRESERVE_CHARS
                            truncated_text = (
                                f"{prefix}...\n\n... [truncated: {removed} chars removed] ...\n\n...{suffix}"
                            )
                            new_items.append({"text": truncated_text})
                            item_changed = True
                            continue

                    new_items.append(item)

                if item_changed:
                    updated_tool_result: Any = {
                        **{k: v for k, v in tool_result.items() if k != "content"},
                        "content": new_items,
                    }
                    new_content.append({"toolResult": updated_tool_result})
                    changes_made = True
                else:
                    new_content.append(content)
                continue

            new_content.append(content)

        if changes_made:
            message["content"] = new_content

        return changes_made

    def _find_oldest_message_with_tool_results(self, messages: Messages) -> int | None:
        """Find the index of the oldest message containing tool results.

        Iterates from oldest to newest so that truncation targets the least-recent
        (and therefore least relevant) tool results first.

        Args:
            messages: The conversation message history.

        Returns:
            Index of the oldest message with tool results, or None if no such message exists.
        """
        # Iterate from oldest to newest
        for idx in range(len(messages)):
            current_message = messages[idx]
            for content in current_message.get("content", []):
                if isinstance(content, dict) and "toolResult" in content:
                    return idx

        return None

File: strands-ts/src/conversation-manager/sliding-window-conversation-manager.ts
/**
 * Sliding window conversation history management.
 *
 * This module provides a sliding window strategy for managing conversation history
 * that preserves tool usage pairs and avoids invalid window states.
 */

import { Message, TextBlock, ToolResultBlock, type ToolResultContent } from '../types/messages.js'
import { DocumentBlock, ImageBlock, VideoBlock } from '../types/media.js'
import type { LocalAgent } from '../types/agent.js'
import { AfterInvocationEvent } from '../hooks/events.js'
import {
  ConversationManager,
  type ProactiveCompressionConfig,
  type ConversationManagerReduceOptions,
} from './conversation-manager.js'
import { logger } from '../logging/logger.js'

const PRESERVE_CHARS = 200
// Max plausible marker length, including newlines. Used as the minimum reduction
// a re-truncation would need to produce in order to be worth running.
const MIN_TRUNCATION_GAIN = 50
// Text payloads at or below this length aren't worth truncating: the savings
// would be smaller than the marker itself, and already-truncated output (which
// lands just above `2 * PRESERVE_CHARS`) falls under this threshold so a
// second pass is a natural no-op.
const TRUNCATION_THRESHOLD = 2 * PRESERVE_CHARS + MIN_TRUNCATION_GAIN

/**
 * Build a short textual stand-in for an image block, used when truncating tool
 * results. The placeholder identifies the image format and its source kind
 * (bytes/url/s3) so the model can reason about what was dropped. For inline
 * bytes the size is included; URL and S3 sources only report the kind since
 * their byte count isn't known locally.
 */
function imagePlaceholder(image: ImageBlock): string {
  const source = image.source
  if (source.type === 'imageSourceBytes') {
    return `[image: ${image.format}, source: bytes, ${source.bytes.byteLength} bytes]`
  }
  if (source.type === 'imageSourceUrl') {
    return `[image: ${image.format}, source: url]`
  }
  return `[image: ${image.format}, source: s3]`
}

/**
 * Build a short textual stand-in for a video block. Binary payloads can't be
 * partially inspected, so videos are replaced wholesale. The placeholder
 * reports format and source kind; byte count is included for inline bytes.
 */
function videoPlaceholder(video: VideoBlock): string {
  const source = video.source
  if (source.type === 'videoSourceBytes') {
    return `[video: ${video.format}, source: bytes, ${source.bytes.byteLength} bytes]`
  }
  return `[video: ${video.format}, source: s3]`
}

/**
 * Build a short textual stand-in for a document block with a binary or remote
 * source. Text-based document sources (text / content) are truncated in place
 * instead of replaced, so this is only called for bytes / s3.
 */
function documentPlaceholder(doc: DocumentBlock): string {
  const source = doc.source
  if (source.type === 'documentSourceBytes') {
    return `[document: ${doc.name}, ${doc.format}, source: bytes, ${source.bytes.byteLength} bytes]`
  }
  return `[document: ${doc.name}, ${doc.format}, source: s3]`
}

/**
 * Build a short textual stand-in for a JSON block. The serialized length is
 * reported so the model knows how much was dropped; truncating JSON
 * mid-structure would produce invalid output, so the whole block is replaced.
 */
function jsonPlaceholder(serializedLength: number): string {
  return `[json: ${serializedLength} chars]`
}

/**
 * Configuration for the sliding window conversation manager.
 */
export type SlidingWindowConversationManagerConfig = {
  /**
   * Maximum number of messages to keep in the conversation history.
   * Defaults to 40 messages.
   */
  windowSize?: number

  /**
   * Whether to truncate tool results when a message is too large for the model's context window.
   * Defaults to true.
   */
  shouldTruncateResults?: boolean

  /**
   * Enable proactive context compression before the model call.
   *
   * - `true`: compress when 70% of the context window is used (default threshold).
   * - `{ compressionThreshold: number }`: compress at the specified ratio (0, 1].
   * - `false` or omitted: disabled, only reactive overflow recovery is used.
   */
  proactiveCompression?: boolean | ProactiveCompressionConfig
}

/**
 * Implements a sliding window strategy for managing conversation history.
 *
 * This class handles the logic of maintaining a conversation window that preserves
 * tool usage pairs and avoids invalid window states. When the message count exceeds
 * the window size, it will either truncate large tool results or remove the oldest
 * messages while ensuring tool use/result pairs remain valid.
 *
 * Registers hooks for:
 * - AfterInvocationEvent: Applies sliding window management after each invocation
 * - AfterModelCallEvent: Reduces context on overflow errors and requests retry (via super)
 * - BeforeModelCallEvent: Proactive compression when threshold is exceeded (via super)
 */
export class SlidingWindowConversationManager extends ConversationManager {
  private readonly _windowSize: number
  private readonly _shouldTruncateResults: boolean

  /**
   * Unique identifier for this conversation manager.
   */
  readonly name = 'strands:sliding-window-conversation-manager'

  /**
   * Initialize the sliding window conversation manager.
   *
   * @param config - Configuration options for the sliding window manager.
   */
  constructor(config?: SlidingWindowConversationManagerConfig) {
    super(config)
    this._windowSize = config?.windowSize ?? 40
    this._shouldTruncateResults = config?.shouldTruncateResults ?? true
  }

  /**
   * Initialize the plugin by registering hooks with the agent.
   *
   * Registers:
   * - AfterInvocationEvent callback to apply sliding window management
   * - AfterModelCallEvent callback to handle context overflow and request retry (via super)
   * - BeforeModelCallEvent callback for proactive compression (via super)
   *
   * @param agent - The agent to register hooks with
   */
  public override initAgent(agent: LocalAgent): void {
    super.initAgent(agent)

    agent.addHook(AfterInvocationEvent, (event) => {
      this._applyManagement(event.agent.messages)
    })
  }

  /**
   * Reduce the conversation history.
   *
   * When `error` is set (reactive overflow recovery), attempts to truncate large tool results
   * first before falling back to message trimming.
   *
   * When `error` is undefined (proactive compression), only trims messages without attempting
   * tool result truncation.
   *
   * @param options - The reduction options
   * @returns `true` if the history was reduced, `false` otherwise
   */
  reduce({ agent, error }: ConversationManagerReduceOptions): boolean {
    return this._reduceContext(agent.messages, error)
  }

  /**
   * Apply the sliding window to the messages array to maintain a manageable history size.
   *
   * Called after every agent invocation. No-op if within the window size.
   *
   * @param messages - The message array to manage. Modified in-place.
   */
  private _applyManagement(messages: Message[]): void {
    if (messages.length <= this._windowSize) {
      return
    }

    this._reduceContext(messages, undefined)
  }

  /**
   * Trim the oldest messages to reduce the conversation context size.
   *
   * The method handles special cases where trimming the messages leads to:
   * - toolResult with no corresponding toolUse
   * - toolUse with no corresponding toolResult
   *
   * The strategy is:
   * 1. First, attempt to truncate large tool results if shouldTruncateResults is true
   * 2. If truncation is not possible or doesn't help, trim oldest messages
   * 3. When trimming, skip invalid trim points (toolResult at start, or toolUse without following toolResult)
   *
   * @param messages - The message array to reduce. Modified in-place.
   * @param _error - The error that triggered the context reduction, if any.
   * @returns `true` if any reduction occurred, `false` otherwise.
   */
  private _reduceContext(messages: Message[], _error?: Error): boolean {
    // Only truncate tool results when handling a context overflow error, not for window size enforcement
    const oldestMessageIdxWithToolResults = this._findOldestMessageWithToolResults(messages)
    if (_error && oldestMessageIdxWithToolResults !== undefined && this._shouldTruncateResults) {
      const resultsTruncated = this._truncateToolResults(messages, oldestMessageIdxWithToolResults)
      if (resultsTruncated) {
        return true
      }
    }

    // Try to trim messages when tool result cannot be truncated anymore
    // If the number of messages is less than the window_size, then we default to 2, otherwise, trim to window size
    let trimIndex = messages.length <= this._windowSize ? 2 : messages.length - this._windowSize

    // Find the next valid trim point that:
    // 1. Starts with a user message (required by some models)
    // 2. Does not start with an orphaned toolResult
    // 3. Does not start with a toolUse unless its toolResult immediately follows
    while (trimIndex < messages.length) {
      const oldestMessage = messages[trimIndex]
      if (!oldestMessage) {
        break
      }

      // Must start with a user message
      if (oldestMessage.role !== 'user') {
        trimIndex++
        continue
      }

      // Cannot start with an orphaned toolResult
      const hasToolResult = oldestMessage.content.some((block) => block.type === 'toolResultBlock')
      if (hasToolResult) {
        trimIndex++
        continue
      }

      // toolUse is only valid if the next message is its toolResult
      const hasToolUse = oldestMessage.content.some((block) => block.type === 'toolUseBlock')
      if (hasToolUse) {
        const nextMessage = messages[trimIndex + 1]
        const nextHasToolResult = nextMessage && nextMessage.content.some((block) => block.type === 'toolResultBlock')
        if (!nextHasToolResult) {
          trimIndex++
          continue
        }
      }

      // Valid trim point found
      break
    }

    // If no valid trim point was found, return false and let the caller handle it.
    // When windowSize is 0, trimIndex === messages.length is expected (remove all), so allow it through.
    if (trimIndex > messages.length || (trimIndex === messages.length && this._windowSize > 0)) {
      logger.warn(
        `window_size=<${this._windowSize}>, messages=<${messages.length}> | unable to trim conversation context, no valid trim point found`
      )
      return false
    }

    // trimIndex is guaranteed to be < messages.length here, so splice always removes at least one message
    messages.splice(0, trimIndex)
    return true
  }

  /**
   * Apply head/tail truncation to a string if it exceeds the size threshold.
   *
   * Returns the truncated form (first {@link PRESERVE_CHARS} + marker + last
   * {@link PRESERVE_CHARS}) when the input exceeds {@link TRUNCATION_THRESHOLD},
   * otherwise `undefined`.
   */
  private _truncateLongText(text: string): string | undefined {
    if (text.length <= TRUNCATION_THRESHOLD) {
      return undefined
    }
    const prefix = text.slice(0, PRESERVE_CHARS)
    const suffix = text.slice(-PRESERVE_CHARS)
    const removed = text.length - 2 * PRESERVE_CHARS
    return `${prefix}\n<truncated chars="${removed}"/>\n${suffix}`
  }

  /**
   * Truncate tool result content in a message to reduce context size.
   *
   * Rule: preserve head/tail when the payload is plain-text-shaped; replace
   * wholesale when it's binary or remote. Specifically:
   * - Text blocks: partial head/tail truncation if over threshold.
   * - Image, Video blocks: wholesale replacement with a textual placeholder.
   * - Document blocks with bytes/s3 source: wholesale replacement.
   * - Document blocks with text source: partial truncation of the inner text.
   * - Document blocks with content source (TextBlock[]): partial truncation of
   *   each nested block.
   * - JSON blocks: wholesale replacement if serialized length is over threshold;
   *   mid-structure truncation would produce invalid JSON.
   *
   * The tool result `status` and `error` fields are preserved.
   *
   * @param messages - The conversation message history.
   * @param msgIdx - Index of the message containing tool results to truncate.
   * @returns True if any changes were made to the message, false otherwise.
   */
  private _truncateToolResults(messages: Message[], msgIdx: number): boolean {
    if (msgIdx >= messages.length || msgIdx < 0) {
      return false
    }

    const message = messages[msgIdx]
    if (!message) {
      return false
    }

    let changesMade = false
    const newContent = message.content.map((block) => {
      if (block.type !== 'toolResultBlock') {
        return block
      }

      const toolResultBlock = block as ToolResultBlock
      const newItems: ToolResultContent[] = []
      let itemChanged = false

      for (const item of toolResultBlock.content) {
        if (item.type === 'imageBlock') {
          newItems.push(new TextBlock(imagePlaceholder(item)))
          itemChanged = true
          continue
        }

        if (item.type === 'videoBlock') {
          newItems.push(new TextBlock(videoPlaceholder(item)))
          itemChanged = true
          continue
        }

        if (item.type === 'documentBlock') {
          const source = item.source
          if (source.type === 'documentSourceBytes' || source.type === 'documentSourceS3Location') {
            newItems.push(new TextBlock(documentPlaceholder(item)))
            itemChanged = true
            continue
          }
          if (source.type === 'documentSourceText') {
            const truncated = this._truncateLongText(source.text)
            if (truncated !== undefined) {
              newItems.push(
                new DocumentBlock({
                  name: item.name,
                  format: item.format,
                  source: { text: truncated },
                  ...(item.citations !== undefined ? { citations: item.citations } : {}),
                  ...(item.context !== undefined ? { context: item.context } : {}),
                })
              )
              itemChanged = true
              continue
            }
          }
          if (source.type === 'documentSourceContentBlock') {
            let nestedChanged = false
            const newContentBlocks = source.content.map((nested) => {
              const truncated = this._truncateLongText(nested.text)
              if (truncated !== undefined) {
                nestedChanged = true
                return new TextBlock(truncated)
              }
              return nested
            })
            if (nestedChanged) {
              newItems.push(
                new DocumentBlock({
                  name: item.name,
                  format: item.format,
                  source: { content: newContentBlocks },
                  ...(item.citations !== undefined ? { citations: item.citations } : {}),
                  ...(item.context !== undefined ? { context: item.context } : {}),
                })
              )
              itemChanged = true
              continue
            }
          }
          newItems.push(item)
          continue
        }

        if (item.type === 'jsonBlock') {
          const serializedLength = JSON.stringify(item.json).length
          if (serializedLength > TRUNCATION_THRESHOLD) {
            newItems.push(new TextBlock(jsonPlaceholder(serializedLength)))
            itemChanged = true
            continue
          }
          newItems.push(item)
          continue
        }

        if (item.type === 'textBlock') {
          const truncated = this._truncateLongText(item.text)
          if (truncated !== undefined) {
            newItems.push(new TextBlock(truncated))
            itemChanged = true
            continue
          }
        }

        newItems.push(item)
      }

      if (!itemChanged) {
        return block
      }

      changesMade = true
      return new ToolResultBlock({
        toolUseId: toolResultBlock.toolUseId,
        status: toolResultBlock.status,
        content: newItems,
        ...(toolResultBlock.error !== undefined ? { error: toolResultBlock.error } : {}),
      })
    })

    if (!changesMade) {
      return false
    }

    messages[msgIdx] = new Message({
      role: message.role,
      content: newContent,
    })

    return true
  }

  /**
   * Find the index of the oldest message containing tool results.
   *
   * Truncation targets the least-recent tool result first so the most relevant
   * recent context is preserved as long as possible.
   *
   * @param messages - The conversation message history.
   * @returns Index of the oldest message with tool results, or undefined if no such message exists.
   */
  private _findOldestMessageWithToolResults(messages: Message[]): number | undefined {
    for (let idx = 0; idx < messages.length; idx++) {
      const currentMessage = messages[idx]!

      const hasToolResult = currentMessage.content.some((block) => block.type === 'toolResultBlock')

      if (hasToolResult) {
        return idx
      }
    }

    return undefined
  }
}
